### ISSUE TITLE
Wrong usage of GOMEMLIMIT causing OOMKilled on istio-proxy

### ISSUE THREAD

**Issue by robert-nemet** [istio/istio#54174]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Hello! We were having issues with OOMKilled pods. After inspection, we discovered it is related to the istio-proxy. I see a GOMEMLIMIT set, but it will limit only istio-proxy, not envoy. What is wrong with a setting like [this](https://github.com/istio/istio/blob/117619e307153461c72502780c0784d47107067b/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml#L277):
```
    - name: GOMEMLIMIT
      valueFrom:
        resourceFieldRef:
          resource: limits.memory
```

The above means that istio-proxy can take all the memory in a container, ignoring other processes, even the envoy. We fixed this by increasing the container's memory limit, but this option should be configurable and not go up to the limit.  I wanted to remove GOMEMLIMIT for istio-proxy, but it is enforced. 
GOMEMLIMIT should be:
- configurable 
- by default, it should only take a small fraction of the limits.memory.

### Version

```Text
istioctl version
client version: 1.24.1
control plane version: 1.23.2
data plane version: 1.23.2 (598 proxies)
----
kubectl version
Client Version: v1.30.5
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.5-gke.1014003
---
helm version --short
v3.16.3+gcfd0749
```

### Additional Information

N/A

**Comment by howardjohn** [comment]
Doesn't GOMEMLIMIT strictly _limit_ the memory and cannot possibly increase it? I agree the limit is essentially "too high" since it doesn't account for Envoy, but its better than no limit?

**Comment by robert-nemet** [comment]
I agree that GOMEMLIMIT is needed. If configured _correctly_, it can prevent OOMKilled. When it is set like this, it can cause issues instead of solving them. Configuration has to take into account other processes, especially envoy. This means istio-proxy(pilot-agent), in this case, has to be a friendly neighbor and give more space for the envoy instead of fighting it.
I wonder if the pilot agent can work stable with a fixed amount of memory or needs a certain portion of the container's memory limit, taking into account the Istio configuration and the number of pods, rules, virtual services, etc.

**Comment by howardjohn** [comment]
How exactly can it cause issues?

Before: it thinks the limit is infinite
After: it thinks the limit is a value that is higher than it should be

It is strictly _limiting_ the memory usage, no?

AFAIK GOMEMLIMIT doesn't mean "Try to reach this memory limit [by making GC less aggressive] and no higher". It means "GC _more_ aggressively if you are near this limit"

**Comment by robert-nemet** [comment]
As already mentioned, the GOMEMLIMIT sets a soft memory limit. The problem is that the sidecar has two main processes: pilot and envoy. The permission given to the pilot to use 100% of available memory needs to be corrected. In my opinion, the pilot should be limited to a smaller portion of `limits.memory`. Actually, as a user, I would like to be able to configure GOMEMLIMIT(from 0 to 100%).

**Comment by howardjohn** [comment]
I don't get it ... 

Without gomemlimit: it can use infinite memory 
With gomemlimit: go will attempt to use at most the limit 

How can adding a limit make it use more memory??

I get we could make it an even lower limit ,but your description is saying it's the cause.

Also,was it even the pilot-agent causing the oom? typically that has extremely low memory usage

**Comment by robert-nemet** [comment]
I think I'm not being understood. I said that today, GOMEMLIMIT is [set like this](https://github.com/istio/istio/blob/117619e307153461c72502780c0784d47107067b/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml#L277):

```yaml
    - name: GOMEMLIMIT
      valueFrom:
        resourceFieldRef:
          resource: limits.memory
```
The setting above tells Go runtime it can count and should try to stay below `limits.memory`. So, for example, if I set `limits.memory` to 200M, the pilot will allow itself to go up to 200Mi(actually, it can go above, but it will try not to). This does not mean it will take everything immediately, but it can count to that memory. Next to the pilot, there is an envoy, which also needs memory.  The envoy's need for a memory is higher than the pilot's. Memory used by the envoy is not visible to the Go runtime. That means it does not count for GOMEMLIMIT. If the memory used by the pilot plus the memory used by the envoy is greater than the `limits.memory` container is killed. Let me know if I'm wrong so far.

GOMEMLIMIT should not be set to 100% of `limits. memory`. Why? From [docs](https://pkg.go.dev/runtime#:~:text=The%20GOMEMLIMIT%20variable,the%20Go%20program.), we see that we need to take into account other processes on the container. 

I'm saying that the option for setting GOMEMLIMIT should be configurable and, by default, should be just a fraction of `limits.memory`.  I'm not saying it should be removed by default. I wanted to remove or edit it while facing issues, but then I learned it is fixed to 100% of `limits.memory`.

**Comment by howardjohn** [comment]
> Let me know if I'm wrong so far.

Yes, that's right. 

What I don't buy is "Wrong usage of GOMEMLIMIT causing OOMKilled". I am fairly confident that, regardless of what GOMEMLIMIT was set to, you would have the same result.


Lets assume the default 1GB memory limit is set. At rest, pilot-agent will generally take ~20mb RSS. Of that, 5mb is heap space.
How Go GC works is, by default, the heap can grow to 2x of its required size before it gets GCed, so we have a range of 5mb - 10mb used by heap, or 20-25mb total. How aggressive this 2x buffer is can be controlled by GOGC and/or GOMEMLIMIT.

So you are OOMing, which implies either:
* Envoy is using >980mb
* pilot-agent is using a lot of memory, which is probably a bug and you should collect a heap dump from localhost:15020/debug/pprof/heap and we can focus on that instead.

Tuning GOMEMLIMIT cannot change this situation. Even if you made Go spend 100% of its time GCing, it would use at most 5mb less.
